src/h3lib/include/alloc.h

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright 2020 Uber Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/** @file alloc.h
+ * @brief   Memory management functions
+ */
+
+#ifndef ALLOC_H
+#define ALLOC_H
+
+#include "h3api.h"
+
+#ifdef H3_ALLOC_PREFIX
+#define H3_MEMORY(name) TJOIN(H3_ALLOC_PREFIX, name)
+
+void* H3_MEMORY(malloc)(size_t size);
+void* H3_MEMORY(calloc)(size_t num, size_t size);
+void* H3_MEMORY(realloc)(void* ptr, size_t size);
+void H3_MEMORY(free)(void* ptr);
```

Comment:
Is there a way to fail this at compile-time if any of these are not defined? Also, how does this work if the custom implementation has also declared these functions in a header file?

Reply:
I don't believe this can be failed at compile time, because the compiler doesn't know that - only the linker does. The replacement memory allocators might only be provided by a host application dynamically.

The custom implementation also declaring these functions is normal (same as sharing the same header file between multiple source files.)